File: Java/src/main/java/OneLang/Test/TestRunner/TestRunner.java
package OneLang.Test.TestRunner;

import io.onelang.std.core.One;
import OneLang.One.CompilerHelper.CompilerHelper;
import OneLang.Test.TestCase.ITestCollection;
import OneLang.Test.TestCase.TestCase;
import OneLang.Test.TestCases.BasicTests.BasicTests;
import OneLang.Test.TestCases.OneFileTests.OneFileTests;
import OneLang.Test.TestCases.ProjectGeneratorTest.ProjectGeneratorTest;

import java.util.List;
import OneLang.Test.TestCase.ITestCollection;
import java.util.Map;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import OneLang.Test.TestCases.BasicTests.BasicTests;
import OneLang.Test.TestCases.OneFileTests.OneFileTests;
import OneLang.Test.TestCases.ProjectGeneratorTest.ProjectGeneratorTest;
import io.onelang.std.core.ExceptionHelper;

public class TestRunner {
    public List<ITestCollection> tests;
    public Map<String, String> argsDict;
    public String outputDir;
    public String baseDir;
    public String[] args;
    
    public TestRunner(String baseDir, String[] args)
    {
        this.baseDir = baseDir;
        this.args = args;
        this.tests = new ArrayList<ITestCollection>();
        this.argsDict = new LinkedHashMap<String, String>();
        CompilerHelper.baseDir = baseDir + "/";
        this.tests.add(new BasicTests());
        this.tests.add(new OneFileTests(this.baseDir));
        this.tests.add(new ProjectGeneratorTest(this.baseDir));
        
        this.parseArgs();
        var getResult = this.argsDict.get("output-dir");
        this.outputDir = (getResult != null ? (getResult) : (baseDir + "/test/artifacts/TestRunner/" + "Java"));
    }
    
    public void parseArgs() {
        for (Integer i = 0; i < this.args.length - 1; i++) {
            if (this.args[i].startsWith("--"))
                this.argsDict.put(this.args[i].substring(2), this.args[i + 1]);
        }
    }
    
    public void runTests() {
        System.out.println("### TestRunner -> START ###");
        
        for (var coll : this.tests) {
            System.out.println("### TestCollection -> " + coll.getName() + " -> START ###");
            for (var test : coll.getTestCases()) {
                System.out.println("### TestCase -> " + test.name + " -> START ###");
                try {
                    var outputDir = this.outputDir + "/" + coll.getName() + "/" + test.name + "/";
                    System.out.println("### TestCase -> " + test.name + " -> OUTPUT-DIR -> " + outputDir + " ###");
                    test.action.accept(outputDir);
                } catch (Exception e)  {
                    System.out.println("### TestCase -> " + test.name + " -> ERROR -> " + ExceptionHelper.toString(e) + " ###");
                }
                System.out.println("### TestCase -> " + test.name + " -> END ###");
            }
            System.out.println("### TestCollection -> " + coll.getName() + " -> END ###");
        }
        
        System.out.println("### TestRunner -> END ###");
    }
}

File: CSharp/Test/TestRunner.cs
using One;
using System.Collections.Generic;
using System.Threading.Tasks;
using System;
using Test.TestCases;
using Test;

namespace Test
{
    public class TestRunner
    {
        public List<ITestCollection> tests;
        public Dictionary<string, string> argsDict;
        public string outputDir;
        public string baseDir;
        public string[] args;
        
        public TestRunner(string baseDir, string[] args)
        {
            this.baseDir = baseDir;
            this.args = args;
            this.tests = new List<ITestCollection>();
            this.argsDict = new Dictionary<string, string> {};
            CompilerHelper.baseDir = $"{baseDir}/";
            this.tests.push(new BasicTests());
            this.tests.push(new OneFileTests(this.baseDir));
            this.tests.push(new ProjectGeneratorTest(this.baseDir));
            
            this.parseArgs();
            this.outputDir = this.argsDict.get("output-dir") ?? $"{baseDir}/test/artifacts/TestRunner/{"CSharp"}";
        }
        
        public void parseArgs()
        {
            for (int i = 0; i < this.args.length() - 1; i++) {
                if (this.args.get(i).startsWith("--"))
                    this.argsDict.set(this.args.get(i).substr(2), this.args.get(i + 1));
            }
        }
        
        public async Task runTests()
        {
            console.log($"### TestRunner -> START ###");
            
            foreach (var coll in this.tests) {
                console.log($"### TestCollection -> {coll.name} -> START ###");
                foreach (var test in coll.getTestCases()) {
                    console.log($"### TestCase -> {test.name} -> START ###");
                    try {
                        var outputDir = $"{this.outputDir}/{coll.name}/{test.name}/";
                        console.log($"### TestCase -> {test.name} -> OUTPUT-DIR -> {outputDir} ###");
                        await test.action(outputDir);
                    } catch (Exception e)  {
                        console.log($"### TestCase -> {test.name} -> ERROR -> {e.ToString()} ###");
                    }
                    console.log($"### TestCase -> {test.name} -> END ###");
                }
                console.log($"### TestCollection -> {coll.name} -> END ###");
            }
            
            console.log($"### TestRunner -> END ###");
        }
    }
}